curated_transformers/layers/attention.py

```diff
@@ -173,23 +174,153 @@ class QkvMode(IntEnum):
+    """
+    ALiBi: Linear biases for attention (`Press et al., 2022`_).
+
+    .. _Press et al., 2022: https://arxiv.org/abs/2108.12409
+    """
+
+    slopes: Tensor
+
+    def __init__(self, *, num_attention_heads: int, is_causal: bool) -> None:
+        """
+        Construct an ALiBi module.
+
+        :param num_attention_heads:
+            Number of attention heads.
+        :param is_causal:
+            Use causal attention.
+        """
+        super().__init__()
+
+        self.is_causal = is_causal
+        slopes = self._calculate_slopes(num_attention_heads)
+        self.register_buffer("slopes", slopes, persistent=False)
+
+    def _calculate_slopes(self, num_attention_heads: int) -> Tensor:
+        """
+        Calculate the linear bias slopes for a given number
+        of attention heads.
+
+        :param num_attention_heads:
+            Number of attention heads.
+        :returns:
+            Head slope tensor.
+
+            *Shape:* ``(1, heads, 1, 1)``
+
+        :meta private:
+        """
+
+        def _slopes_with_step(num_attention_heads, *, step=1):
+            ratio = 2.0 ** (-8.0 / num_attention_heads)
+            return ratio ** torch.arange(1, 1 + num_attention_heads, step)
+
+        # The slope as proposed in the ALiBi paper would be:
+        #
+        # return _slopes_with_step(num_attention_heads)
+        #
+        # However the authors note in their implementation that using powers
+        # of 2 for n in the ratio 2**(-8/n) of the geometric sequence for
+        # slopes has better properties.
+        #
+        # Most implementations use powers use powers of two in the following
+        # manner: if the number of heads is not a power of 2, then we find
+        # k=the largest power of 2 in 1..n. The slopes are then computed
+        # as the concatenation of:
+        #
+        # - The slopes for 1..k.
+        # - The slopes for 1..2*k with step 2, takin the first n-k elements.
+
+        # k is the largest power of 2 in 1..n.
+        k = 1 << ((num_attention_heads).bit_length() - 1)
+        slopes = _slopes_with_step(k)
+
+        if num_attention_heads != k:
+            remaining_heads = num_attention_heads - k
+            slopes_rest = _slopes_with_step(2 * k, step=2)[:remaining_heads]
+            slopes = torch.cat([slopes, slopes_rest])
+
+        return slopes.view(1, -1, 1, 1)
+
+    def _calculate_biases(self, slopes: Tensor, seq_len: int) -> Tensor:
+        """
+        Calculate the linear bias tensor upto a given (key) sequence length.
+
+        :param slopes:
+            Slope tensor.
+
+            *Shape:* ``(1, heads, 1, 1)``
+        :param seq_len:
+            Maximum number of timesteps to calculate.
+        :returns:
+            Multi-headed linear bias tensor.
+
+            *Shape:* ``(1, heads, seq_len, seq_len)`` (non-causal) or
+            ``(1, heads, 1, seq_len)`` (causal)
+
+        :meta private:
+        """
+        if not self.is_causal:
```

Comment:
Can we flip the branches around, then we can have `if self.is_causal:`

Reply:
Done.